trunner/device.py

```diff
@@ -374,13 +389,52 @@ def reset(self):
         time.sleep(0.050)
         self.reset_gpio.high()
 
-    def boot(self, serial_downloader=False):
+    def restart(self):
+        # without powering down rpi ports eval kit is partially powered through usb
+        rpi_ports_off = subprocess.run([
+            'uhubctl',
+            '-l', '2',
+            '-a', '0'],
+            stdout=subprocess.DEVNULL
+        )
+        if rpi_ports_off.returncode != 0:
+            logging.error('uhubctl failed!\n')
+            raise Exception('RPi ports power down failed!')
+
+        self.restart_gpio.low()
+        time.sleep(0.500)
+        self.reset_gpio.low()
+        self.restart_gpio.high()
+        time.sleep(0.500)
+        self.reset_gpio.high()
+        rpi_ports_on = subprocess.run([
+            'uhubctl',
+            '-l', '2',
+            '-a', '1'],
+            stdout=subprocess.DEVNULL
+        )
+        if rpi_ports_on.returncode != 0:
+            logging.error('uhubctl failed!\n')
+            raise Exception('RPi ports power up failed!')
+
+        # after power on there is pre-restart device, which will disappear
+        wait_for_disappear(DEVICE_SERIAL, timeout=5)
+        try:
+            wait_for_dev(DEVICE_SERIAL, timeout=5)
+        except TimeoutError:
+            logging.error('Serial port not found!\n')
+            sys.exit(1)
+
+    def boot(self, serial_downloader=False, restart=True):
         if serial_downloader:
             self.boot_gpio.low()
         else:
             self.boot_gpio.high()
 
-        self.reset()
+        if restart:
+            self.restart()
+        else:
+            self.reset()
 
     def flash(self):
         self.boot(serial_downloader=True)
```

Comment:
Why we restart a board before each test but not before flashing?

Reply:
We restart a board before flashing in self.boot(serial_downloader=True) after 'def flash(self):'